distribution/lib/Standard/Base/0.0.0-dev/src/Data/Decimal.enso

```diff
@@ -706,13 +707,21 @@ type Decimal
        Conversion to `Integer` uses truncation: the fractional portion is
        dropped.
 
+       If the conversion resulted in a loss of precision, a
+       `Loss_Of_Numeric_Precision` warning is attached.
+
        > Example
          Convert `Decimal` 2345.6 to an `Integer`.
              d = Decimal.new "2345.6"
              d.to_integer
              # => 2345
     to_integer : Integer
-    to_integer self = self.big_decimal.toBigInteger
+    to_integer self =
+        i = if self <= 9223372036854775807 && self >= -9223372036854775808 then self.big_decimal.longValue else
```

Comment:
This shall not be necessary. I suggest to:
```diff
diff --git distribution/lib/Standard/Base/0.0.0-dev/src/Data/Decimal.enso distribution/lib/Standard/Base/0.0.0-dev/src/Data/Decimal.enso
index 128ac5f791..f1d48b5573 100644
--- distribution/lib/Standard/Base/0.0.0-dev/src/Data/Decimal.enso
+++ distribution/lib/Standard/Base/0.0.0-dev/src/Data/Decimal.enso
@@ -717,8 +717,7 @@ type Decimal
              # => 2345
     to_integer : Integer
     to_integer self =
-        i = if self <= 9223372036854775807 && self >= -9223372036854775808 then self.big_decimal.longValue else
-            self.big_decimal.toBigInteger
+        i = self.big_decimal.toBigInteger
         #IO.println "HMM "+self.to_text+" "+i.to_text+" "+(self == i).to_text
         if self == i then i else
             Warning.attach (Loss_Of_Numeric_Precision.Warning self i) i
diff --git engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeNode.java engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeNode.java
index f933f8d7a9..add1204977 100644
--- engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeNode.java
+++ engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeNode.java
@@ -143,12 +143,16 @@ public abstract class HashCodeNode extends Node {
     return bigInteger.getValue().hashCode();
   }
 
-  @Specialization(guards = {"!interop.fitsInLong(v)", "interop.fitsInBigInteger(v)"})
+  @Specialization(guards = {"interop.isNumber(v)"})
   @TruffleBoundary
   long hashCodeForBigInteger(
       Object v, @Shared("interop") @CachedLibrary(limit = "10") InteropLibrary interop) {
     try {
-      return interop.asBigInteger(v).hashCode();
+      if (interop.fitsInLong(v)) {
+        return hashCodeForLong(interop.asLong(v));
+      } else {
+        return interop.asBigInteger(v).hashCode();
+      }
     } catch (UnsupportedMessageException ex) {
       throw EnsoContext.get(this).raiseAssertionPanic(this, "Expecting BigInteger", ex);
     }
```
CCing @Akirathan

Reply:
This works -- thank you @JaroslavTulach.